include/deal.II/meshworker/copy_data.h

```diff
@@ -0,0 +1,155 @@
+#include <deal.II/lac/vector.h>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace MeshWorker
+{
+  /**
+   * Helper copy data struct.
+   *
+   * This class is a good default drop in CopyData object for the
+   * WorkStream::run() and MeshWorker::mesh_loop() functions.
+   *
+   * It allows you to store arbitrary number of (local) full matrices, vectors,
+   * and local degrees of freedom index vectors.
+   *
+   * @author Luca Heltai, 2019.
+   */
+  template <int n_matrices    = 1,
+            int n_vectors     = n_matrices,
+            int n_dof_indices = n_matrices>
+  struct CopyData
+  {
+    /**
+     * Initialize everything with the same @p size. This is usually the number
+     * of local degrees of freedom.
+     */
+    CopyData(const unsigned int size);
+
+    /**
+     * For every object, specify the size they should have.
+     */
+    CopyData(
+      const std::array<std::array<unsigned int, 2>, n_matrices> &matrix_sizes,
+      const std::array<unsigned int, n_vectors> &                vector_sizes,
+      const std::array<unsigned int, n_dof_indices> &dof_indices_sizes);
+
+    /**
+     * Deep copy constructor.
+     */
+    CopyData(const CopyData<n_matrices, n_vectors, n_dof_indices> &other) =
+      default;
+
+    /**
+     * Allow resetting of all elements of the struct to zero, by simply
+     * calling `(*this) = 0;`
+     *
+     * Notice that the only allowed number here is really `0`. Calling this
+     * function with any other number will result in an Assertion.
+     *
+     * The elements of the arrays of local degrees of freedom indices are
+     * all set to numbers::invalid_dof_index.
+     */
+    void
+    operator=(const double &number);
+
+    /**
+     * An array of local matrices.
+     */
+    std::array<FullMatrix<double>, n_matrices> matrices;
+
+    /**
+     * An array of local vectors.
+     */
+    std::array<Vector<double>, n_vectors> vectors;
+
+    /**
+     * An array of local degrees of freedom indices.
+     */
+    std::array<std::vector<types::global_dof_index>, n_dof_indices>
+      local_dof_indices;
+  };
+
+
+#ifndef DOXYGEN
+  //
+  // Template definitions
+  //
+  template <int n_matrices, int n_vectors, int n_dof_indices>
+  CopyData<n_matrices, n_vectors, n_dof_indices>::CopyData(
+    const unsigned int size)
+  {
+    for (auto &m : matrices)
+      m.reinit({size, size});
+    for (auto &v : vectors)
+      v.reinit(size);
+    for (auto &d : local_dof_indices)
+      d.resize(size);
+  }
+
+
+
+  template <int n_matrices, int n_vectors, int n_dof_indices>
+  CopyData<n_matrices, n_vectors, n_dof_indices>::CopyData(
+    const std::array<std::array<unsigned int, 2>, n_matrices> &matrix_sizes,
+    const std::array<unsigned int, n_vectors> &                vector_sizes,
+    const std::array<unsigned int, n_dof_indices> &dof_indices_sizes)
+  {
+    unsigned int i = 0;
+    for (auto &m : matrices)
+      m.reinit(matrix_sizes[i++]);
```

Comment:
I find this idiom hard to read -- this may be a case where it's easier to write a traditional `for` loop because then it is clear 
 what objects you are comparing.

Same of course in the lines below.